Config flow for the Flux integration

schelv opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
I've made some key improvements to the Flux integration:

**Improvements:**
1. **Simplified Installation:** Flux can now be installed and set up directly from the UI, removing the need for yaml configuration.
2. **UI Parameter Consolidation:** I've consolidated `brightness` and `disable_brightness_adjust` into one UI option, "Adjust Brightness".

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/27742

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

schelv committed:
flux integration config flow

schelv committed:
small changes to settings

schelv committed:
adapted tests for config flow

schelv committed:
Merge branch 'home-assistant:dev' into flux-config-flow

schelv committed:
tiny cleanup

schelv committed:
remove commented code

schelv committed:
update manifest and codeowners

schelv committed:
add requirements and dependencies to manifest

schelv committed:
re-added yaml config code

schelv committed:
removed another piece of code in comments

schelv committed:
Tried to improve the UI text

schelv committed:
Converted test case written by @JakeHillion

schelv committed:
rename file

schelv committed:
remove commented code

schelv committed:
remove UNDEFINED values in config flow

schelv committed:
add config_flow.py tests

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'home-assistant:dev' into flux-config-flow

schelv committed:
Merge branch 'home-assistant:dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Update homeassistant/components/flux/__init__.py

jpbede commented on homeassistant/components/flux/config_flow.py:
```diff
@@ -0,0 +1,232 @@
+    async def async_step_user(self, user_input=None) -> FlowResult:
+        """Handle the initial step."""
+
+        if self._async_current_entries():
+            return self.async_abort(reason="already_configured")
```
This prevents the creation of multiple flux entries, which should IMO be possible.

schelv replied:
The main reason why I added this restriction is because I couldn't think of a reason why someone would want to have multiple flux configurations at the same time. 

It's a pretty weak reason haha.
I will make some changes and add it back.

Having multiple flux configurations could have a problem I think.
What happens when a lamp is added to multiple configurations?
Will both configurations attempt to control the light?
This sounds like something that's unwanted.

But this would be an existing problem and not one introduced by this pull request.
Still it's something to think about.

jpbede commented on homeassistant/components/flux/config_flow.py:
```diff
@@ -0,0 +1,232 @@
+
+        if user_input is not None:
+            user_input.update(default_settings())
+            return self.async_create_entry(title="Flux", data=user_input)
```
`title="Flux"` should be something unique or let the user input the name for it.

jpbede commented on homeassistant/components/flux/config_flow.py:
```diff
@@ -0,0 +1,232 @@
+    return settings_dict
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Flux."""
```
The config flow is missing `async_step_import` to import existing configurations in `configuration.yaml`.

jpbede commented on homeassistant/components/flux/switch.py:
```diff
@@ -193,12 +226,49 @@ def __init__(
+
+    async_add_entities([flux])
+
+    platform = entity_platform.async_get_current_platform()
+
+    async def async_update() -> None:
+        """Update lights."""
+        await flux.async_flux_update()
+
+    platform.async_register_entity_service(
+        "flux_update",
+        {},
+        async_update,
+    )
```
Can be removed, when migrated to config entries.

schelv replied:
@jpbede is this comment still relevant? If so can you explain why this can be removed?

jpbede replied:
Ok, nvm, this should be kept. Thought it would be a reload service, to update the config :)

schelv replied:
As far as I can see all comments are resolved now. Can you take a second look at the code?

jpbede commented on homeassistant/components/flux/switch.py:
```diff
@@ -45,29 +46,34 @@
-MODE_RGB = "rgb"
-DEFAULT_MODE = MODE_XY
 
 PLATFORM_SCHEMA = vol.Schema(
     {
         vol.Required(CONF_PLATFORM): "flux",
         vol.Required(CONF_LIGHTS): cv.entity_ids,
         vol.Optional(CONF_NAME, default="Flux"): cv.string,
         vol.Optional(CONF_START_TIME): cv.time,
+        vol.Optional(CONF_SUNSET_TIME): cv.time,
         vol.Optional(CONF_STOP_TIME): cv.time,
         vol.Optional(CONF_START_CT, default=4000): vol.All(
             vol.Coerce(int), vol.Range(min=1000, max=40000)
```
Can be removed, when migrated to config entries.

jpbede commented on homeassistant/components/flux/manifest.json:
```diff
@@ -2,8 +2,12 @@
+  "config_flow": true,
+  "dependencies": [],
   "documentation": "https://www.home-assistant.io/integrations/flux",
+  "integration_type": "hub",
```
```suggestion
  "integration_type": "service",
```

jpbede commented on homeassistant/components/flux/switch.py:
```diff
@@ -193,12 +226,49 @@ def __init__(
+class FluxSwitch(SwitchEntity, RestoreEntity):
+    """Representation of a Flux switch."""
+
+    def __init__(
+        self,
+        name: str,
+        hass: HomeAssistant,
+        lights,
+        start_time: time | None,
+        sunset_time: time | None,
+        stop_time: time | None,
+        start_colortemp,
+        sunset_colortemp,
+        stop_colortemp,
+        brightness,
+        disable_brightness_adjust,
+        mode,
+        interval: timedelta,
+        transition: timedelta,
+    ) -> None:
```
Maybe a dataclass would make sense?

schelv replied:
The use of a dataclass does seem logical. However, I question if this merge request is the right moment to introduce it. The second tip on [this page](https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr) advises against making small unrelated improvements.

emontnemery replied:
A dataclass makes sense, but it should indeed happen in a follow-up PR, not in this PR 👍

jpbede commented on homeassistant/components/flux/__init__.py:
Consider adding "async_setup" which creates a repair declaring that yaml has been migrated to config entries. E.g. https://github.com/home-assistant/core/blob/23d5fb962211f95cce9dc34458fa321752399b82/homeassistant/components/zodiac/__init__.py#L21

jpbede replied:
Please see https://github.com/home-assistant/architecture/blob/master/adr/0021-YAML-integration-configuration-deprecation-policy.md also

jpbede reviewed (changes requested):
First of all, thanks a lot for your contribution! Thanks 👍 

To start the review, some comments. I just took a quick look :)

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Update homeassistant/components/flux/manifest.json

Co-authored-by: Jan-Philipp Benecke <github@bnck.me>

schelv committed:
support multiple flux config entries

schelv committed:
remove unneeded test

schelv committed:
Not adding new feature to yaml config

schelv committed:
Adding warning about yaml config

schelv committed:
Merge branch 'dev' into flux-config-flow

jpbede commented:
At https://github.com/home-assistant/core/pull/95447 you can see how a migration to the configuration flow works.

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv commented:
Sorry for my late response.
I think it might be better to have the conversion of the yaml configuration in a separate pull request.
I'm not sure if or when I will implement it.
Maybe someone wants to help implement the next step.
Is there a place where I can create an "issue" for the work that needs to be done?
The "Issues" of this project is for bugs only for some reason

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
remove introduction of new option

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
fix yaml to config entry migration

schelv committed:
remove unreachable statement

schelv committed:
add test for migration of option

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv commented:
@jpbede I've added the migration to the configuration flow. Thanks again for the review!

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv commented:
@jpbede  Can you take another look at the code?

schelv committed:
Merge branch 'dev' into flux-config-flow

joostlek commented on homeassistant/components/flux/config_flow.py:
```diff
@@ -0,0 +1,268 @@
+        return self.async_create_entry(title=entry_data[CONF_NAME], data=entry_data)
+
+
+class OptionsFlow(config_entries.OptionsFlow):
+    """Handle flux options."""
+
+    def __init__(self, config_entry: config_entries.ConfigEntry) -> None:
+        """Initialize the flux options flow."""
+        self._config_entry = config_entry
```
```suggestion
class FluxOptionsFlow(OptionsFlowWithConfigEntry):
    """Handle flux options."""
```

joostlek replied:
We have a class for that :D

joostlek commented on homeassistant/components/flux/config_flow.py:
```diff
@@ -0,0 +1,268 @@
+
+        if user_input is not None:
+            user_input.update(default_settings())
+            return self.async_create_entry(title=user_input[CONF_NAME], data=user_input)
```
mutable data should be stored in the options instead of the data field

joostlek commented on homeassistant/components/flux/config_flow.py:
```diff
@@ -0,0 +1,268 @@
+        """Get the options flow for the Flux component."""
+        return OptionsFlow(config_entry)
+
+    async def async_step_user(self, user_input=None) -> FlowResult:
```
```suggestion
    async def async_step_user(self, user_input: dict[str, Any] | None=None) -> FlowResult:
```

joostlek commented on homeassistant/components/flux/config_flow.py:
```diff
@@ -0,0 +1,268 @@
+        if CONF_STOP_TIME in yaml_config:
+            entry_data[CONF_STOP_TIME] = str(yaml_config.pop(CONF_STOP_TIME))
+
+        yaml_config.pop("platform")
```
You don't really have to pop everything, you just get the data you need and that's it

joostlek commented on homeassistant/components/flux/config_flow.py:
```diff
@@ -0,0 +1,268 @@
+        """Initialize the flux options flow."""
+        self._config_entry = config_entry
+
+    def reset_values_to_default(self, user_input):
```
Why do we want to use the default again?

schelv replied:
After using non-default values, there should be a way to use the default values again.
For example after configuring a sunset time, you decide you want to use the sunset time based on location and time of the year again.
With the yaml config this could be done simply by removing a key value pair from the yaml file.
With the GUI I could not find a simple way to "unset" a value, so I used this hack.

Most default values are static but the "default" values related to (sunset) time change during the year.

schelv replied:
Finally found out how to make the value clearable🥳

emontnemery replied:
@schelv What did you do to make the values clearable?

schelv replied:
I changed
```
vol.Optional(
                  CONF_START_TIME,
                  default=settings.get(CONF_START_TIME, UNDEFINED),
              ): TimeSelector(),
```
to 
```
vol.Optional(
                  CONF_START_TIME,
                  description={"suggested_value": settings.get(CONF_START_TIME)},
              ): TimeSelector(),
```

and proposed [this frontend merge request](https://github.com/home-assistant/frontend/pull/18590)

emontnemery replied:
I see. Does it mean this PR depends on a merge of that frontend PR? If so, please link to it in the PR description and add tag `awaiting-frontend`

schelv replied:
As far as I remember it is possible to clear the value without the frontend merge request. Not very easily though.

joostlek commented on homeassistant/components/flux/config_flow.py:
```diff
@@ -0,0 +1,268 @@
+
+    async def async_step_init(self, user_input=None) -> FlowResult:
+        """Configure the options."""
+        errors: dict[str, str] = {}
```
You're not really raising any errors

joostlek commented on homeassistant/components/flux/manifest.json:
```diff
@@ -2,8 +2,12 @@
-  "codeowners": [],
+  "codeowners": ["@schelv"],
+  "config_flow": true,
+  "dependencies": [],
```
Remove empty objects

schelv committed:
Apply suggestions from code review

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
make two defaults constant

schelv committed:
apply suggestions

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
fix

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
start and stop time can now be cleared

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
config entry tests not needed because of migration

emontnemery commented on homeassistant/components/flux/config_flow.py:
```diff
@@ -0,0 +1,255 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+MINIMAL_FLUX_SCHEMA = vol.Schema(
```
```suggestion
USER_FLOW_SCHEMA = vol.Schema(
```

emontnemery commented on homeassistant/components/flux/config_flow.py:
```diff
@@ -0,0 +1,255 @@
+    """Handle flux options."""
+
+    def convert_mired_to_kelvin(self, user_input):
+        """Convert between mireds and kelvins because I can't find the kelvin option for ColorTempSelector."""
```
```suggestion
        """Convert between mireds and kelvins ColorTempSelector only supports mireds."""
```

emontnemery commented on homeassistant/components/flux/strings.json:
```diff
@@ -0,0 +1,37 @@
+        "data": {
+          "name": "A name for your flux configuration",
+          "lights": "Select the lights you want controlled by Flux",
+          "start_time": "Start time. If left empty this will be automatically calculated based on the location.",
```
This, and several of the others, are too long and chatty for a label, please split between `data` and `data_description`. Also, if *automatically calculated based on the location* means it's calculated according to sunset and sunrise, please write that instead.

emontnemery commented on homeassistant/components/flux/switch.py:
```diff
@@ -137,19 +151,52 @@ async def async_setup_platform(
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Flux lights."""
+    name = entry.options.get(CONF_NAME, entry.title)
```
Why do we `get` from the options, aren't these all unconditionally set?

schelv replied:
The original code used `get`'s. I guess I just kept them.

schelv replied:
Should I change this?

emontnemery replied:
The original code is also wrong since CONF_NAME defaults to "Flux". Please change to not use get 👍

schelv committed:
Apply suggestions from code review

Co-authored-by: Erik Montnemery <erik@montnemery.com>

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
apply emontnemery's suggestions

schelv committed:
add some data descriptions

schelv committed:
not using get

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
use new color temp selector

joostlek commented:
Bumping this one to be on my list again

emontnemery commented:
I discussed this PR with @balloob , @frenck  and some other core developers. The concern is that the `flux` integration has known issues and limitations and we don't want to promote it by adding a config flow to it at this point.

@balloob thinks the functionality implemented by flux is very useful, but he suggested a deeper integration with Home Assistant than just an integration. Maybe a built-in option for lights with adjustable color-temperature to make them automatically adjust the color temperature during the day?

Another view (which I personally do not share) is that the flux integration is too simple and could just be replaced with a blueprint, such as this one https://community.home-assistant.io/t/advanced-circadian-lighting/383543

There's also a custom integration with more options: https://github.com/claytonjn/hass-circadian_lighting/

Until the way forward is clear, we do not want to merge this PR.

schelv commented:
@emontnemery Thank you for the comment.

In an attempt to turn this reply into a ~~concise and~~ coherent story, I've chosen to structure it as follows:
I will start by providing some context; Explain what in my view the goal should be, and what the most feasible approach is to reach it.
Then I will reply/address/discuss the concern and alternatives that were mentioned.

**Some context + chosen approach**
My opinion is that light control (including the functionality provided by flux) should be easy to use by users of all knowledge levels. (If installation is required, it should also be easy)
It currently is not easy to configure/use this in home assistant.

I know that it can be very difficult to make something "easy to use".
I do have (big) ideas on how things can be improved, but do not have the knowledge and experience required to effectively implement them. (and getting it merged might also be a challenge 😉).
Contributing to this project becomes more realistic if I make an existing thing "easier to use" with multiple smaller changes.
 
That is why I decided to improve this integration.
In its current state it probably isn't good, but:
- it has potential to be easily installable (because it is in the core),
- by iteratively adding multiple small improvements, it might even become decent. 

The first small improvement I had in mind was this one: https://github.com/home-assistant/core/pull/93181
But I found out I was the third person to propose exactly this change.

Apparently making yaml changes is not allowed, and a config flow is needed. (See https://github.com/home-assistant/core/pull/86201#discussion_r1080679662)
That is the reason why this merge request exists.


**Suggested alternatives.**

I will address/discuss alternatives that were mentioned, and relate them to the context provided above.

> There are also a couple of well maintained and popular custom integration with more options:

The integrations seem really good, but sadly they need to be installed using HACS.
HACS itself is not easy to install by users of _all_ knowledge levels. ("clear your browser caches", having/creating a Github account)


> @balloob ... suggested a deeper integration with Home Assistant than an integration. ... Going in this direction would require an architecture discussion.

As stated earlier I don't feel I have the knowledge to implement architecture changes.
But I do like the idea of deeper integration. Not having to install/add an integration is definitely very easy.
Maybe the deeper integration could be a (very future) continuation of flux (without the issues and limitations).

**The Concern**

> I discussed this PR with @balloob , @frenck and some other core developers. The concern is that the `flux` integration has known issues and limitations and we don't want to promote it by adding a config flow to it at this point.

I don't know what the "known issues and limitations" of `flux` are.
But I can understand that you don't want to promote an integration if it has issues.

Proposed way forward:
- Someone makes a list of the "known issues and limitations" (knowing them is a prerequisite).
- Identify which issues and limitations need to be addressed before `flux` becomes config-flow-worthy.
- Someone addresses the issues and limitations that need to be addressed before `flux` becomes config-flow-worthy.
- `flux` becomes config-flow-worthy

schelv committed:
Merge branch 'dev' into flux-config-flow

schelv committed:
attempt to fix checks

frenck commented:
Hi there @schelv

As pointed out correctly above, we indeed don't want to move forward with this at this point.

You ask about the issues with this integration; there are quite a few (especially when working with multiple light sources), but also feature completeness.

From as user perspective point of view, eventually, we don't want this to be an integration that provides a "device automation"-as-integration, but a native, built-in part of just lights in Home Assistant.

In the end, wrapping a light in an integration to automate stuff like this is odd. For that same reason, we don't want to level up a few more of these integrations (like alert and device_sun_light_trigger).

Nevertheless, thanks for being willing to contribute ❤️ 

../Frenck
